### ISSUE TITLE
[Feature request]:  provide proper no sandbox mode for IDE tooling that includes program access

### ISSUE THREAD

**Issue by ell1e** [flatpak/flatpak#5644]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a feature request that matches the one I want to file, without success.

### Suggestion

*(My apologies if this is a duplicate or if it's already available, but given all code editors I've ever tried on FlatHub other than maybe GNOME Builder suffer from this problem, it doesn't seem to be resolved)*

**The problem:**

As is elaborated here https://bugs.kde.org/show_bug.cgi?id=431223 and in so many other places I've come across, whatever is found on FlatHub of **most dev tooling seems to be currently borderline unusable unless you work on the most standard project types,** due to the sandboxing. This happens even with full host filesystem access enabled (which it usually is for IDE tools) which then actually leaves arguably little to no security benefit at this point anyway, because the environment vars are too different and separate, and as a result no tooling the user may have installed outside seems to be usually accessible. This can be seen [here with what is apparently needed to work around that](https://gitlab.gnome.org/GNOME/gnome-builder/-/commit/fee5214c927c841b66cfd34c0321b8c9890149ce).

This seems to be especially a problem for: 1. unusual LSPs (no flatpak extension image stuff can ever solve that because there's always another one that nobody thought of, it doesn't scale), 2. unusual extra tools people may want to use in the IDE terminal like different style checkers or formatters, 3. different build tools people may want to use in the IDE terminal that aren't the usual ones used for that programming language, 4. probably more I didn't think of.

**Solutions that don't really work for many use cases:**

1. [Language extensions.](https://github.com/flathub?q=org.freedesktop.Sdk.Extension&type=all&language=&sort=) These fundamentally don't work well because 1. niche languages won't have these, 2. niche tooling won't be in these, if I want just a single more unusual tool in my build process it'll just not be there, 3. no IDE I've ever seen works around this in a manner I think would be understandable for newcomers, e.g. by having some integrated automatically interactively suggested way of installing such missing tools then and guiding the user how to do it. Instead they usually just fall flat on their face and say command not found with zero explanation of why or how to even install more system level packages via apt or such, let alone even explain what system package manager is used inside the sandbox.

2. `flatpak-spawn`: rather hostile to beginners, someone who tries an internal IDE terminal isn't going to know how this works. And again, no IDE that I've ever seen explains this via e.g. some note appearing in the terminal once I try to run a missing command that is detected to exist on the host. Even if it did, if e.g. my project contains a prewritten build script that expects to directly call a lot of stuff without flatpak-spawn that is installed in various non-trivial ways on my host it's just not going to work without major pain.

**A potential solution:**

As a solution, I propose that there's a proper, extended no sandboxing mode that will 1. merge PATH such that maybe the stuff inside the flatpak filesystem is secondary, and everything from the outside PATH is preferred, 2. set and use all the XDG config, cache, socket, whatever dirs from the outside system including /tmp and such, 3. whatever else I may have forgotten that is needed to make most of the "outside" tooling work. Basically, what [gnome builder is currently manually doing, or similar.](https://gitlab.gnome.org/GNOME/gnome-builder/-/commit/fee5214c927c841b66cfd34c0321b8c9890149ce)

I'm aware this also introduces new corner case problems due to the merged PATH and whatnot, and possibly binaries from the inside unintentionally calling things from the outside instead of intended neighboring inside copies. However, a lot of this can be avoided when the inside image of the flatpak contains for example a proper portable install of the app, e.g. by wrapping an AppImage.

I'm aware this making-it-portable-if-it-wasn't-to-start-with flatpak was originally also intended to solve which such a new mode would kind of in parts undo, but the thing is, flatpak also solves a lot of other things, among them 1. way more accessible methods of distribution for beginners than figuring out how to download & find again an AppImage in your filesystem and manually keep that up to date and all that, 2. better directories and methods of browsing of what is even available via platforms like FlatHub, 3. better update mechanisms to rely on for the app developer if the app doesn't bother with an integrated one, 4. integrates better into immutable-style distros, 5. other stuff I probably didn't think of.

All of these advantages would remain available, so IDE tooling which wanted to access these but wanted to go the extra mile of wrapping a more portable binary inside the flatpak (which they might already have at zero cost if they also provide AppImages) could then use this new mode to make sure that the user can actually access all dev tooling they would ever want with mostly no more weird separations and indirections.

**TL;DR:**

I suggest flatpak lets IDE stuff trade some portability and risk more clashes between the binaries and such, via a even-less-sandboxing mode, in favor of actually accessing tooling on the host system properly without many restrictions. (Without manually needing to figure out and do the hackery that gnome builder is doing.) While this puts some more work on the IDE authors, it's imho arguably very little work for IDEs that already have a portable mode, way more flatpak-independent work, and way easier to do than to somehow integrate UI elements to explain flatpak-spawn to an IDE user which still wouldn't solve all issues with this anyway.

**Comment by TingPing** [comment]
I'm not sure I completely understand what you propose but I don't think there is any possibility of improving how it work much.

When an app is in a sandbox it sees the sandbox, there is no way to transparently also see the host, or mix in the host.

So an app has to be aware of different environments that it runs in. In 2023 I think this is a basic function of an IDE.

**Comment by ell1e** [comment]
If this is a "basic function", then why does almost no IDE on flathub, including by the way VS Code and VS Codium, handle it properly?

**Comment by TingPing** [comment]
VSCode has native support for running in different environments:

- https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-ssh
- https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-wsl

Nobody wrote support for Flatpak specifically though.

**Comment by TingPing** [comment]
GNOME-Builder uses Flatpak's Spawn DBus APIs to run things on the host, then from the host it can run inside of different sandboxes (flatpak, podman, toolbx, etc).

**Comment by ell1e** [comment]
The commit I linked doesn't really look like such a clean separation, more like a mixed environment for a terminal. That's also what I think most IDEs would benefit from as a possible default option (picked on a per flatpak sandbox option).

**Comment by TingPing** [comment]
It is a completely clean separation, it just copies around some environment variables.

**Comment by ell1e** [comment]
> It is a completely clean separation, it just copies around some environment variables.

I don't understand, aren't these copied for the purpose of removing the isolation? E.g. it seems to set things to not only the host XDG dirs but also the host dbus, and even host wayland and x11 sockets and display info, from what I can see. That doesn't see isolated at all, which seems to me like what I was proposing. Basically, there should be a mode where accessing the host programs and graphical env and all is possible out of the box in cases where effective sandboxing isn't desired.

**Comment by smcv** [comment]
> Basically, there should be a mode where accessing the host programs and graphical env and all is possible out of the box

This is not possible within the design of Flatpak. Flatpak programs are allowed to expect and require `/usr` to be the Flatpak runtime, but host programs expect and require `/usr` to be the host OS. Even in the absence of any attempt at sandboxing or security, there would be no execution environment where both of those categories of programs get what they expect.

If your IDE cannot run like this, then Flatpak might not be the right approach to running your IDE: it does not necessarily have to be the right way to distribute and run every possible program in the world.

> when the inside image of the flatpak contains for example a proper portable install of the app, e.g. by wrapping an AppImage

It seems you have designed a different app framework that is not Flatpak. You're very welcome to create or use that framework, but Flatpak isn't it.

A "proper portable install" is not something that is possible to do without making various assumptions: for instance AppImage assumes that the host system's glibc, OpenGL drivers and various other libraries (I don't know whether the exact list is documented anywhere) are "close enough" to being a drop-in replacement for the versions of the corresponding libraries that the app was originally compiled against. The design of Flatpak aims to avoid needing to make those assumptions.

**Comment by ell1e** [comment]
> If your IDE cannot run like this

It seems like most can't, from all those that I've tried. A portable IDE should work if `/usr` was just the one of the host OS, hence me proposing that mode as an opt-in. And yes this is only the case for glibc IDEs on a glibc host, but you could still whitelist it only for that case. The glibc version doesn't really need to match [since you can build most software, at least outside of the latest C++, in a way that works with a wide range of glibc versions](https://stackoverflow.com/a/62998901/1204141).

I just think almost no IDE properly working is a bit of a problem. I guess GNOME Builder might work I honestly don't know, but I haven't seen any other that doesn't completely break down if you just try to use custom language servers, often even the standard ones, inside flatpak, let alone if you try to actually build the project.

**Comment by TingPing** [comment]
I'm going to close this issue because what you want simply isn't what Flatpak is; The things you discuss are not trivial or close to how it currently functions.

An IDE will *never* behave identical to a host application inside of the sandbox, because it is inside of the sandbox.

IDEs need to be aware of this detail and supporting running things in different contexts. GNOME-Builder is the primary example of this.

**Comment by ell1e** [comment]
I see! In that case I would call into question if having any of the not-GNOME-Builder-IDEs on flathub even makes sense at all. I guess that would be best to bring up with the flathub people?

That IDEs need to be aware of this seems at least right now seems a bit like wishful thinking. But as a result, it just makes it seem like the whole "Develop" section on flathub is kind of mostly broken while not advertised as such, which doesn't really reflect well for the usefulness of it, or does it? I wouldn't know either of any flathub-like registry for all distributions for dev tooling releases to fill the gap instead. So this seems like an ecosystem problem.
